### PR TITLE
HTCONDOR-1484 fix for eval crash by adding to_delete ptrs to EvalState

### PR DESCRIPTION
**Pull request by johnkn**
and modifying EvaluateExpr and EvaluateAttr to convert results to self-contained by default. This changes the classad ABI, so the SOVERSION is updated to 16

<Insert PR description here, and leave checklist below for code review.>

# HTCondor Pull Request Checklist for internal reviewers

- [ ] Verify that (GitHub thinks) the merge is clean. If it isn't, and you're confident you can resolve the conflicts, do so. Otherwise, send it back to the original developer.
- [ ] Verify that the related Jira ticket exists and has a target version number and that it is correct.
- [ ] Verify that the Jira ticket is in review status and is assigned to the reviewer.
- [ ] Verify that the Jira ticket (HTCONDOR-xxx) is mentioned at the beginning of the title. Edit it, if not
- [ ] Verify that the branch destination of the PR matches the target version of the ticket
- [ ] Check for correctness of change
- [ ] Check for regression test(s) of new features and bugfixes (if the feature doesn't require root)
- [ ] Check for documentation, if needed
- [ ] Check for version history, if needed
- [ ] Check BaTLab dashboard for successful build (https://batlab.chtc.wisc.edu/results/workspace.php) and test for either the PR or a workspace build by the developer that has the Jira ticket as a comment.
- [ ] Check that each commit message references the Jira ticket (HTCONDOR-xxx)

## After the above
- Hit the merge button if the pull request is approved and it is not a security patch (security changes require 2 additional reviews)
- If the pull request is approved, take the ticket out of review state
- Assign JIRA Ticket back to the developer

### CODE REVIEW

**Comment by timtheisen** [comment]
This change is insufficient. Corresponding packaging changes need to be done. Maybe we can remove all the SOVERSION stuff since we we really do not want to support a standalone ClassAd library.

File: src/classad/exprTree.cpp [hunk 1]
```diff
@@ -353,6 +331,49 @@ SetRootScope( )
...
+	return false;
+}
+
+// if a Value currently has a pointer that is also in the evaluation cache
+// transfer ownership of the pointer to the Value (removing it from the cache)
+// and convert the Value to a counted-pointer type, thus making the value self-contained.
+// This is an optimization for the case where evaluation creates a temporary ExprTree
+// that is the result of evaluation.  We can avoid deep-copying the tree and then
+// turning around and deleting the original.
+bool EvalState::GivePtrToValue(Value & val)
```
Comment (JaimeFrey) [inline src/classad/exprTree.cpp:361]: This feels like it should just be in the body of Value::MakeSelfContained(). Then, Value doesn't have to make EvalState a friend.

**Review by JaimeFrey — CHANGES_REQUESTED** [review]
All notes are inline.

**Comment by JaimeFrey** [comment]
I don't see any mention of the ClassAds library SO version in the files under build/packaging/, just wildcards on the library filename.

File: src/classad/exprTree.cpp [hunk 2]
```diff
@@ -353,6 +331,49 @@ SetRootScope( )
...
+		cache_to_delete.push_back(tree);
+	}
+}
+
+// look for the given ExprTree pointer in the cache, and if it is found
+// remove it from the cache and return true.
+bool EvalState::TakeFromDeletionCache(ExprTree * tree) {
+	auto last = std::remove(cache_to_delete.begin(), cache_to_delete.end(), tree);
+	if (last != cache_to_delete.end()) {
+		cache_to_delete.erase(last, cache_to_delete.end());
```
[the remark this answers is not in the document]
Comment (johnkn) [reply src/classad/exprTree.cpp:349]: This code is from Greg.  remove guarantees that the iterator points to the first of the elements to be removed, or to the first element after the end if it is not found.

[the remark this answers is not in the document]
Comment (JaimeFrey) [reply src/classad/exprTree.cpp:349]: The two-argument call to erase() you have removes the matching element and all elements after it. We only want to erase the single element, which is what the single-argument form does.

[the remark this answers is not in the document]
Comment (JaimeFrey) [reply src/classad/exprTree.cpp:349]: Oops. I missed that the preceding call is a std::remove(), and not a find() like in the other function. I retract my suggestion.

**Review by JaimeFrey — APPROVED** [review]
Looks good now.

**Pull request merged**